libsplinter/src/service/scabbard/client/upload.rs

```diff
@@ -0,0 +1,413 @@
+        load_smart_contract_from_file(&scar_file_path)
+    }
+
+    /// Attempt to convert the smart contract definition into a `SabrePayload`.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if the Sabre `Action` or `SabrePayload` builders fail.
+    pub fn try_into_sabre_payload(self) -> Result<SabrePayload, Error> {
+        let create_contract = CreateContractActionBuilder::new()
+            .with_name(self.metadata.name)
+            .with_version(self.metadata.version)
+            .with_inputs(self.metadata.inputs)
+            .with_outputs(self.metadata.outputs)
+            .with_contract(self.contract)
+            .build()
+            .map_err(|err| {
+                Error::new_with_source("failed to build CreateContractAction", err.into())
+            })?;
+
+        let payload = SabrePayloadBuilder::new()
+            .with_action(Action::CreateContract(create_contract))
+            .build()?;
+
+        Ok(payload)
+    }
+}
+
+/// The metadata of a smart contract that needs to be included in the Sabre transaction.
+#[derive(Debug, Deserialize, Serialize)]
+pub struct SmartContractMetadata {
+    name: String,
+    version: String,
+    inputs: Vec<String>,
+    outputs: Vec<String>,
+}
+
+fn determine_scar_file_path(scar: &str) -> Result<PathBuf, Error> {
+    if arg_is_file_path(scar) {
+        Ok(PathBuf::from(scar))
+    } else {
+        let scar_paths = var_os(SCAR_PATH_ENV_VAR).ok_or_else(|| {
+            Error::new(&format!(
+                "cannot find .scar file: {} not set",
+                SCAR_PATH_ENV_VAR
+            ))
+        })?;
+        split_paths(&scar_paths)
+            .find_map(|mut path| {
+                path.push(scar);
+                if path.exists() {
+                    Some(path)
+                } else {
+                    path.set_extension(SCAR_FILE_EXTENSION);
+                    if path.exists() {
+                        Some(path)
+                    } else {
+                        None
+                    }
+                }
+            })
+            .ok_or_else(|| Error::new(&format!("{} not found in {}", scar, SCAR_PATH_ENV_VAR)))
+    }
+}
+
+fn arg_is_file_path(arg: &str) -> bool {
+    arg.contains('/')
+}
+
+fn load_smart_contract_from_file(file_path: &Path) -> Result<SmartContract, Error> {
+    let scar_file = File::open(file_path).map_err(|err| {
+        Error::new_with_source(
+            &format!("failed to open file {}", file_path.display()),
+            err.into(),
+        )
+    })?;
+    let mut archive = Archive::new(BzDecoder::new(scar_file));
+    let archive_entries = archive
+        .entries()
+        .map_err(|err| Error::new_with_source("failed to read .scar file", err.into()))?;
+
+    let mut metadata = None;
+    let mut contract = None;
+
+    for entry in archive_entries {
+        let mut entry = entry.map_err(|err| {
+            Error::new_with_source("invalid .scar: failed to read archive entry", err.into())
```

Comment:
instead of "invalid .scar file:", suggest "invalid scar file:"

Reply:
Updated